### ISSUE TITLE
Using clang is impure

### ISSUE THREAD

**Issue by nh2** [NixOS/nixpkgs#24697]
## Issue description

Clang has default linker search paths looking for `.so` files in `/usr/lib`. That makes builds impure and even fail if it picks up a library from `/usr/lib` that dynamically links a library to somewhere not in the default linker paths.

It probably has the same problem for the include path.

### Steps to reproduce

`strings /nix/store/path/to/bin/clang | grep /usr/lib`

## Possible solutions

* Patch clang or provide a configure-time option to get rid of these default paths (they are not present in the `gcc` binary either)
* Run clang with the `--sysroot=/nonexistent` argument which will make it not use the default search paths.

Note that I couldn't find `--sysroot` in `man clang` or `clang --help` (but it works and is documented on https://clang.llvm.org/docs/CrossCompilation.html); I could only find `-isysroot` in `--help`, which is likely the solution to the corresponding default include path problem.

**Comment by copumpkin** [comment]
Hmm, I think our wrapper does the right thing for `clang` because we use it as the primary compiler for everything on Darwin and it's not reaching out to anything impure. Might still be better to patch the actual binary though.

**Comment by nh2** [comment]
@copumpkin Clang doesn't use the same default search paths on Linux and Darwin.

Check the code for Linux here: https://github.com/llvm-mirror/clang/blob/be15c6a8274d96827ac707e5075118d4c376940f/lib/Driver/ToolChains.cpp#L4272

```c++
  addPathIfExists(D, SysRoot + "/lib", Paths);
  addPathIfExists(D, SysRoot + "/usr/lib", Paths);
```

**Comment by dtzWill** [comment]
Sorry, I think I missed this:
1) I use llvm/clang extensively BUT with a custom wrapper for various reasons.
2) I actually go the `--sysroot` route, although IIRC clang gets unhappy if you point it at invalid paths, instead it must be a valid directory but contain nothing useful for it to use :/.  I also set a few more things when trying to free clang of its strong desire to find and imitate the local gcc:
```nix
''...
${optionalString (!needsGCC) ''
  ccBothFlags=" --sysroot=$cc"
  ccCFlags+=" -ccc-gcc-name $cc" # invalid
  ccCFlags+=" --gcc-toolchain=$cc" # also invalid
  # XXX: Do this elsewhere!
  ccLDFlags+=" --enable-new-dtags"
''}
...''
```

Basically I point clang at itself :P.

I think I've seen commits since 4.0 that improve Clang's ability to operate standalone on Linux without excessive kludgery, but I haven't been following that closely.  There's a proposal bounced around on and off about supporting use of a "config" file which would be *great* for our use cases IMO.

Anyway, what should we do to fix this? Patch out those paths, or maybe set sysroot as you suggest?

**Comment by nh2** [comment]
> Anyway, what should we do to fix this? Patch out those paths, or maybe set sysroot as you suggest?

@dtzWill From my perspective both seem OK, patching it out or using flags; whichever is easier / more reliable / more maintainable.

**Comment by nh2** [comment]
It's difficult though. What should happen if you use a nix-built `clang` in a `nix-shell` on Ubuntu?

It may be desirable that it _can_ pick up libraries from `/usr/lib` in general (so you can use it like a normal `clang` on Ubuntu), but when in a `nix-shell`, it'll probably be undesirable that it picks those up instead of libraries you provide in your `nix-shell`. To get reliable, desired behaviour, you'd have to sandbox the `nix-shell` too.

That's why I'd argue these tools work best in "no-baked-in-paths" mode. `lld` already follows this concept already apparently.

**Comment by Twey** [comment]
@matthewbauer ‘I don't think we want to handicap our compilers’ — but the linker wrapper script already explicitly checks for `/usr/lib` &c. references and errors out, so the only thing that will come of the compiler picking up impure libraries is errors.  We've generally made a point in Nix and nixpkgs to try to enforce purity, or at least explicitly flag impurity; are we changing that policy now we have sandboxing?  We still don't have sandboxing everywhere (e.g. `nix-shell`).

**Comment by Twey** [comment]
Bump.

**Comment by orivej** [comment]
For the reference, I have noticed that `clang --sysroot=/path` does not look for headers in `/path/usr/include` (as it usually does) since this change: https://github.com/NixOS/nixpkgs/commit/507d65d7d4f195ecd5b08ee5d24f6e96924f7b1c

**Comment by nh2** [comment]
I've just run into this again:

Using `clang` together with `meson` in `nix-shell` on Ubuntu with an opencv dependency made meson determine that it should pass files like `/usr/lib/x86_64-linux-gnu/libopencv_highgui.so` to the linker invocation.

**Comment by matthewbauer** [comment]
Still important

**Comment by Twey** [comment]
Still relevant

**Comment by Twey** [comment]
@matthewbauer We just hit this yet again at my company, and I still don't have a good solution to offer.  Is there any clarification as to whether we want to continue eliminating references to global paths from packages, or whether we're okay with relying on sandboxing for isolation?

**Comment by tobim** [comment]
This should probably be patched. We can add a hook to re-add the upstream default search paths if people really want that, but I doubt there is a use-case.

**Comment by RossComputerGuy** [comment]
Is this issue still prevalent? I just checked Clang in LLVM 18 myself and I don't see any references to `/usr`.

**Comment by nh2** [comment]
Would be good if somebody could find out _what_ fixed it, to be sure about it.

**Comment by RossComputerGuy** [comment]
Well, it probably had to do with the `gnu-install-dirs.patch` we have. We probably could close this issue now as it is fixed.
